diff -urNp -x '*.orig' util-linux-2.36.2.org/login-utils/su-common.c util-linux-2.36.2/login-utils/su-common.c
--- util-linux-2.36.2.org/login-utils/su-common.c	2021-02-12 12:32:01.749988723 +0100
+++ util-linux-2.36.2/login-utils/su-common.c	2021-02-28 14:22:27.508126765 +0100
@@ -100,6 +100,15 @@ UL_DEBUG_DEFINE_MASKNAMES(su) = UL_DEBUG
 
 #define is_pam_failure(_rc)	((_rc) != PAM_SUCCESS)
 
+/* The default paths which get set are both bogus and oddly influenced
+    by <paths.h> and -D on the commands line. Just to be clear, we'll set
+    these explicitly. -ewt */
+#undef DEFAULT_LOGIN_PATH
+#undef DEFAULT_ROOT_LOGIN_PATH
+#define DEFAULT_LOGIN_PATH "/bin:/usr/bin:/usr/local/bin:/usr/X11R6/bin"
+#define DEFAULT_ROOT_LOGIN_PATH \
+    "/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/usr/X11R6/bin:/usr/X11R6/sbin"
+	
 /* The shell to run if none is given in the user's passwd entry.  */
 #define DEFAULT_SHELL "/bin/sh"
 
