diff -urN util-linux-2.12.org/fdisk/llseek.c util-linux-2.12/fdisk/llseek.c
--- util-linux-2.12.org/fdisk/llseek.c	2003-11-16 15:32:12.000000000 +0100
+++ util-linux-2.12/fdisk/llseek.c	2003-11-16 15:32:36.000000000 +0100
@@ -19,7 +19,7 @@
 
 #else	/* HAVE_LLSEEK */
 
-#if defined(__alpha__) || defined(__ia64__) || defined(__s390x__)
+#if defined(__alpha__) || defined(__ia64__) || defined(__s390x__) || defined(__x86_64__)
 
 #define my_llseek lseek
 
diff -urN util-linux-2.12.org/misc-utils/whereis.c util-linux-2.12/misc-utils/whereis.c
--- util-linux-2.12.org/misc-utils/whereis.c	2003-11-16 15:32:12.000000000 +0100
+++ util-linux-2.12/misc-utils/whereis.c	2003-11-16 15:32:43.000000000 +0100
@@ -66,6 +66,8 @@
    "/usr/etc",
    "/lib",
    "/usr/lib",
+   "/lib64",
+   "/usr/lib64",
    "/usr/games",
    "/usr/games/bin",
    "/usr/games/lib",
