diff -ru util-linux-2.11b/mount/Makefile util-linux-2.11b-fixed/mount/Makefile
--- util-linux-2.11b/mount/Makefile	Mon Apr  2 19:09:46 2001
+++ util-linux-2.11b-fixed/mount/Makefile	Mon Apr  2 19:09:26 2001
@@ -57,7 +57,7 @@
 main_losetup.o: lomount.c rmd160.o
 	$(COMPILE) -DMAIN lomount.c -o $@
 
-losetup: main_losetup.o $(LIB)/xstrncpy.o
+losetup: $(LO_OBJS)
 	$(LINK) $^ -o $@
 
 mount.o umount.o nfsmount.o losetup.o fstab.o realpath.o sundries.o: sundries.h
