--- util-linux-2.13-pre7/mount/umount.c.orig	2007-04-16 17:01:33.000000000 +0200
+++ util-linux-2.13-pre7/mount/umount.c	2007-04-16 17:01:43.000000000 +0200
@@ -21,7 +21,7 @@
 #include "env.h"
 #include "nls.h"
 
-#if defined(MNT_FORCE) && !defined(__sparc__) && !defined(__arm__)
+#if defined(MNT_FORCE)
 /* Interesting ... it seems libc knows about MNT_FORCE and presumably
    about umount2 as well -- need not do anything */
 #else /* MNT_FORCE */
--- util-linux-2.13-pre7/schedutils/ionice.c.orig	2007-04-16 17:30:44.000000000 +0200
+++ util-linux-2.13-pre7/schedutils/ionice.c	2007-04-16 17:33:38.000000000 +0200
@@ -33,6 +33,9 @@
 #elif defined(__s390x__) || defined(__s390__)
 #define __NR_ioprio_set		282
 #define __NR_ioprio_get		283
+#elif defined(__sparc__)
+#define __NR_ioprio_set		196
+#define __NR_ioprio_get		218
 #else
 #error "Unsupported arch"
 #endif
