--- util-linux-2.9v/MCONFIG.MCONFIG	Tue Jul 20 02:53:55 1999
+++ util-linux-2.9v/MCONFIG	Tue Jul 27 11:34:38 1999
@@ -16,17 +16,17 @@
 # If HAVE_PAM is set to "yes", then login, chfn, chsh, and newgrp
 # will use PAM for authentication. Additionally, passwd will not be
 # installed as it is not PAM aware.
-HAVE_PAM=no
+HAVE_PAM=yes
 
 # If HAVE_SHADOW is set to "yes", then login, chfn, chsh, newgrp, passwd,
 # and vipw will not be built or installed from the login-utils
 # subdirectory.  
-HAVE_SHADOW=yes
+HAVE_SHADOW=no
 
 # If HAVE_PASSWD is set to "yes", then passwd will not be built or
 # installed from the login-utils subdirectory (but login, chfn, chsh,
 # newgrp, and vipw *will* be installed).
-HAVE_PASSWD=no
+HAVE_PASSWD=yes
 
 # If you use chfn and chsh from this package, REQUIRE_PASSWORD will require
 # non-root users to enter the account password before updating /etc/passwd.
@@ -72,7 +72,7 @@
 # If HAVE_SLN is set to "yes", then sln won't be installed
 # (but the man page sln.8 will be installed anyway).
 # sln also comes with libc and glibc.
-HAVE_SLN=no
+HAVE_SLN=yes
 
 CC=		gcc
 
@@ -111,7 +111,7 @@
 # SLANGFLAGS=-I/usr/include/slang
 # No such subdirectory - slcurses.h lives in /usr/include
 #   (no extra definition required).
-LIBSLANG=-lslang
+# LIBSLANG=-lslang
 
 
 DEVDIR=		$(DESTDIR)/dev
@@ -124,13 +124,13 @@
 BINDIR=         $(DESTDIR)/bin
 VARPATH=	$(DESTDIR)/var
 LOGDIR=		$(DESTDIR)/var/log
-MANDIR=		$(DESTDIR)/usr/man
+MANDIR=		$(DESTDIR)/usr/share/man
 MAN1DIR=	$(MANDIR)/man1
 MAN3DIR=	$(MANDIR)/man3
 MAN5DIR=	$(MANDIR)/man5
 MAN6DIR=	$(MANDIR)/man6
 MAN8DIR=	$(MANDIR)/man8
-INFODIR=	$(DESTDIR)/usr/info
+INFODIR=	$(DESTDIR)/usr/share/info
 
 # Directory for shutdown, halt, reboot, etc.
 SHUTDOWNDIR=	$(SBINDIR)
@@ -147,7 +147,7 @@
 MANMODE=	644
 DATMODE=	644
 INFOMODE=	644
-SUIDMODE=      4755
+SUIDMODE=      4711
 
 CHMOD=		chmod
 INSTALL=	install
@@ -155,7 +155,7 @@
 INSTALLBIN=	$(INSTALL) -m $(BINMODE)
 INSTALLMAN=	$(INSTALL) -m $(MANMODE)
 INSTALLDAT=	$(INSTALL) -m $(DATMODE)
-INSTALLSUID=    $(INSTALL) -m $(SUIDMODE) -o root
+INSTALLSUID=    $(INSTALL) -m $(SUIDMODE)
 
 %.o: %.c
 	$(CC) -c $(CFLAGS) $< -o $@
