--- util-linux-2.10k/sys-utils/ipcs.c.glibc	Sun Nov 21 18:36:03 1999
+++ util-linux-2.10k/sys-utils/ipcs.c	Mon Apr 17 05:07:05 2000
@@ -31,6 +31,8 @@
 /* X/OPEN tells us to use <sys/{types,ipc,sem}.h> for semctl() */
 /* X/OPEN tells us to use <sys/{types,ipc,msg}.h> for msgctl() */
 /* X/OPEN tells us to use <sys/{types,ipc,shm}.h> for shmctl() */
+#define _GNU_SOURCE 1
+#define __USE_GNU 1
 #include <sys/types.h>
 #include <sys/ipc.h>
 #include <sys/sem.h>
