--- util-linux-2.13-pre7/misc-utils/Makefile.am~	2007-03-24 01:14:12.000000000 +0100
+++ util-linux-2.13-pre7/misc-utils/Makefile.am	2007-03-24 01:14:27.000000000 +0100
@@ -35,7 +35,8 @@
 if HAVE_NCURSES
 usrbinexec_PROGRAMS += setterm
 man_MANS += setterm.1
-setterm_LDADD = -lncurses
+setterm_LDADD = -lncurses -ltinfo
+setterm_CFLAGS = -I/usr/include/ncurses
 endif
 
 if BUILD_KILL
--- util-linux-2.13-pre7/text-utils/Makefile.am.orig	2005-08-14 13:33:02.000000000 +0200
+++ util-linux-2.13-pre7/text-utils/Makefile.am	2007-03-24 01:21:01.000000000 +0100
@@ -9,9 +9,12 @@
 if HAVE_NCURSES
 bin_PROGRAMS = more
 usrbinexec_PROGRAMS += ul pg
-more_LDADD = -lncurses
+more_LDADD = -lncurses -ltinfo
+more_CFLAGS = -I/usr/include/ncurses
 pg_LDADD = -lncurses
-ul_LDADD = -lncurses
+pg_CFLAGS = -I/usr/include/ncurses
+ul_LDADD = -lncurses -ltinfo
+ul_CFLAGS = -I/usr/include/ncurses
 man_MANS += ul.1 more.1 pg.1
 else
 if HAVE_TERMCAP
