--- util-linux-2.12/MCONFIG~	2004-05-23 12:26:51.792803744 +0200
+++ util-linux-2.12/MCONFIG	2004-05-23 12:32:31.114219048 +0200
@@ -48,13 +48,13 @@
 # be built or installed from the login-utils subdirectory.  (The shutdown
 # and halt that come with the SysVinit package should be used with the init
 # found in that package.)
-HAVE_SYSVINIT=yes
+HAVE_SYSVINIT=no
 
 # If HAVE_SYSVINIT_UTILS is set to "yes", then last, mesg, and wall will
 # not be built or installed from the login-utils subdirectory.  (The
 # shutdown and init from the SysVinit package do not depend on the last,
 # mesg, and wall from that package.)
-HAVE_SYSVINIT_UTILS=yes
+HAVE_SYSVINIT_UTILS=no
 
 # If HAVE_WRITE is set to "yes", then write will not be built or
 # installed from the misc-utils subdirectory.
