diff -urNp -x '*.orig' util-linux-2.36.2.org/disk-utils/blockdev.c util-linux-2.36.2/disk-utils/blockdev.c
--- util-linux-2.36.2.org/disk-utils/blockdev.c	2021-02-02 16:41:37.962357499 +0100
+++ util-linux-2.36.2/disk-utils/blockdev.c	2021-02-28 14:22:07.717508138 +0100
@@ -431,6 +431,7 @@ static void report_all_devices(void)
 	char ptname[200 + 1];
 	char device[210];
 	int ma, mi, sz;
+	char iobuf[32*1024];
 
 	procpt = fopen(_PATH_PROC_PARTITIONS, "r");
 	if (!procpt)
@@ -444,6 +445,7 @@ static void report_all_devices(void)
 		snprintf(device, sizeof(device), "/dev/%s", ptname);
 		report_device(device, 1);
 	}
+	setvbuf(procpt, iobuf, _IOFBF, sizeof(iobuf));
 
 	fclose(procpt);
 }
diff -urNp -x '*.orig' util-linux-2.36.2.org/disk-utils/fdisk-list.c util-linux-2.36.2/disk-utils/fdisk-list.c
--- util-linux-2.36.2.org/disk-utils/fdisk-list.c	2021-02-12 12:32:01.733988782 +0100
+++ util-linux-2.36.2/disk-utils/fdisk-list.c	2021-02-28 14:22:07.717508138 +0100
@@ -320,6 +320,7 @@ done:
 char *next_proc_partition(FILE **f)
 {
 	char line[128 + 1];
+	static char iobuf[32*1024];
 
 	if (!*f) {
 		*f = fopen(_PATH_PROC_PARTITIONS, "r");
@@ -327,6 +328,7 @@ char *next_proc_partition(FILE **f)
 			warn(_("cannot open %s"), _PATH_PROC_PARTITIONS);
 			return NULL;
 		}
+		setvbuf(*f, iobuf, _IOFBF, sizeof(iobuf));
 	}
 
 	while (fgets(line, sizeof(line), *f)) {
