diff -urN util-linux-2.11m.org/configure util-linux-2.11m/configure
--- util-linux-2.11m.org/configure	Wed Feb 20 13:22:58 2002
+++ util-linux-2.11m/configure	Wed Feb 20 13:24:32 2002
@@ -431,7 +431,7 @@
 #
 # 7. Does xgettext exist and take the option --foreign-user?
 #
-if (test $ENABLE_NLS = yes && type xgettext > /dev/null 2>&1); then
+if (test $ENABLE_NLS = yes && test -x /usr/bin/xgettext > /dev/null 2>&1); then
 	msg=`xgettext --foreign-user 2>&1 | grep unrecognized`
 	if test -n "$msg"; then
 		echo "FOREIGN = " >> make_include
