--- ./disk-utils/raw.c.orig	Sat Jun 30 13:03:56 2001
+++ ./disk-utils/raw.c	Sat Jun 30 13:54:54 2001
@@ -19,7 +19,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <sys/sysmacros.h>
-#include <linux/raw.h>
+#include <sys/raw.h>
 #include <linux/major.h>
 
 #ifdef OLD_RAW_DEVS
@@ -31,6 +31,8 @@
 #endif
 #define DEVFS_RAWCTLDEV "/dev/raw/rawctl"
 
+#undef __LINUX_RAW_H
+#define __LINUX_RAW_H __SYSTEM_RAW_H
 
 char *	progname;
 int	do_query = 0;

--- util-linux-2.10q.org/configure      Tue Nov 21 22:08:46 2000
+++ util-linux-2.10q/configure  Tue Nov 21 22:10:35 2000
@@ -33,6 +33,7 @@
 # 12. For hwclock.c: does struct tm have a field tm_gmtoff?
 # 13. For nfsmount: does the output of rpcgen compile?
 # 14. For fsck.cramfs, mkfs.cramfs: do we have libz?
+# 15. For raw.c: does <sys/raw.h> exist?

 rm -f make_include defines.h

@@ -680,4 +681,11 @@
         echo "You don't have zlib"
 fi
 rm -f conftest conftest.c
+
+#
+# 15. For raw.c: does <sys/raw.h> exist?
+#
+if ./testincl "sys/raw.h"; then
+       echo "ADD_RAW=yes" >> make_include
+fi
 
