diff -ur util-linux-2.11z.old/fdisk/Makefile util-linux-2.11z/fdisk/Makefile
--- util-linux-2.11z.old/fdisk/Makefile	2003-06-14 18:05:54.000000000 +0000
+++ util-linux-2.11z/fdisk/Makefile	2003-06-14 18:06:10.000000000 +0000
@@ -21,10 +21,8 @@
 SBIN:=$(SBIN) fdisk
 MAN8:=$(MAN8) fdisk.8
 
-ifneq "$(ARCH)" "sparc"
 SBIN:=$(SBIN) $(CFDISK) sfdisk
 MAN8:=$(MAN8) cfdisk.8 sfdisk.8
-endif
 
 all: $(SBIN) $(NOTMADE)
 
Tylko w util-linux-2.11z/fdisk: Makefile~
