--- util-linux-2.20.1/configure.ac~	2011-11-27 21:36:59.000000000 +0100
+++ util-linux-2.20.1/configure.ac	2011-11-27 21:44:54.089830949 +0100
@@ -491,7 +491,7 @@
   AC_DEFINE([HAVE_SCANF_AS_MODIFIER], [1], [scanf %as modifier]) ;;
 *)
   AC_MSG_RESULT([no])
-  if "x$build_libmount" = xyes; then
+  if test "x$build_libmount" = xyes; then
     AC_MSG_WARN([%as or %ms for sscanf() not found; do not build libmount])
     build_libmount=no
   fi
