--- util-linux-2.10m/mount/nfsmount.c.nfsv3	Thu Jun 22 11:24:15 2000
+++ util-linux-2.10m/mount/nfsmount.c	Thu Jun 22 11:25:14 2000
@@ -106,7 +106,7 @@ find_kernel_nfs_mount_version(void) {
 	if (kernel_version) {
 	     if (kernel_version < MAKE_VERSION(2,1,32))
 		  nfs_mount_version = 1;
-	     else if (kernel_version < MAKE_VERSION(2,3,99))
+	     else if (kernel_version < MAKE_VERSION(2,2,16))
 		  nfs_mount_version = 3;
 	     else
 		  nfs_mount_version = 4; /* since 2.3.99pre4 */
