diff -urN util-linux-2.11b.org/mount/nfsmount.c util-linux-2.11b/mount/nfsmount.c
--- util-linux-2.11b.org/mount/nfsmount.c	Wed Mar 21 10:07:38 2001
+++ util-linux-2.11b/mount/nfsmount.c	Wed Mar 21 10:13:29 2001
@@ -724,6 +724,16 @@
 		server_addr.sin_port = PMAPPORT;
 		port = pmap_getport(&server_addr, nfsprog, nfsvers,
 				    tcp ? IPPROTO_TCP : IPPROTO_UDP);
+		/* Here we check to see if user is mounting with the tcp
+		 * option.  If so, and if the portmap returns a '0' for
+		 * port (service unavailable), we then exit out notifying
+		 * the user, rather than hanging up mount. */
+
+		if (port == 0 && tcp == 1) {
+			perror(_("nfs server reported service unavailable"));
+			goto fail;
+		}
+
 		if (port == 0)
 			port = NFS_PORT;
 #ifdef NFS_MOUNT_DEBUG
