diff -Nur util-linux-2.9s/text-utils/Makefile util-linux-2.9s.pld/text-utils/Makefile
--- util-linux-2.9s/text-utils/Makefile	Fri Jun 11 08:39:14 1999
+++ util-linux-2.9s.pld/text-utils/Makefile	Fri Jun 11 08:38:57 1999
@@ -20,7 +20,7 @@
 
 # Where to put datebase files?
 
-USRLIB=		more.help
+DATAMISC=		more.help
 
 all: $(BIN) $(USRBIN)
 
@@ -68,10 +68,10 @@
 endif
 
 install install.shadow install.text-utils: all
-	$(INSTALLDIR) $(BINDIR) $(USRBINDIR) $(USRLIBDIR) $(MAN1DIR)
+	$(INSTALLDIR) $(BINDIR) $(USRBINDIR) $(DATAMISCDIR) $(MAN1DIR)
 	$(INSTALLBIN) $(BIN) $(BINDIR)
 	$(INSTALLBIN) $(USRBIN) $(USRBINDIR)
-	$(INSTALLDAT) $(USRLIB) $(USRLIBDIR)
+	$(INSTALLDAT) $(DATAMISC) $(DATAMISCDIR)
 	$(INSTALLMAN) $(MAN1) $(MAN1DIR)
 
 .PHONY: clean distclean
