--- ./disk-utils/raw.c.orig	Sat Jun 30 13:03:56 2001
+++ ./disk-utils/raw.c	Sat Jun 30 13:54:54 2001
@@ -19,7 +19,7 @@
 #include <sys/stat.h>
 #include <sys/ioctl.h>
 #include <sys/sysmacros.h>
-#include <linux/raw.h>
+#include <sys/raw.h>
 #include <linux/major.h>
 
 #ifdef OLD_RAW_DEVS
@@ -30,6 +30,8 @@
 #define RAWDEVDIR "/dev/raw/"
 #endif
 
+#undef __LINUX_RAW_H
+#define __LINUX_RAW_H __SYSTEM_RAW_H
 
 char *	progname;
 int	do_query = 0;
