diff -urNp -x '*.orig' util-linux-2.36.2.org/include/pathnames.h util-linux-2.36.2/include/pathnames.h
--- util-linux-2.36.2.org/include/pathnames.h	2021-02-12 12:32:01.736988771 +0100
+++ util-linux-2.36.2/include/pathnames.h	2021-02-28 14:21:31.543044005 +0100
@@ -39,7 +39,7 @@
 #define _PATH_NOLOGIN_TXT	"/etc/nologin.txt"
 
 #ifndef _PATH_MAILDIR
-# define _PATH_MAILDIR		"/var/spool/mail"
+# define _PATH_MAILDIR		"/var/mail"
 #endif
 #define	_PATH_MOTDFILE		"/usr/share/misc/motd:/run/motd:/etc/motd"
 #ifndef _PATH_NOLOGIN
--- util-linux-2.37/login-utils/login.1.adoc.orig	2021-05-06 12:55:46.912423336 +0200
+++ util-linux-2.37/login-utils/login.1.adoc	2021-07-04 11:07:54.188129931 +0200
@@ -141,7 +141,7 @@
 _/var/run/utmp_,
 _/var/log/wtmp_,
 _/var/log/lastlog_,
-_/var/spool/mail/*_,
+_/var/mail/*_,
 _/etc/motd_,
 _/etc/passwd_,
 _/etc/nologin_,
