--- util-linux-2.37/sys-utils/lscpu-cputype.c.orig	2021-05-26 13:32:10.618108196 +0200
+++ util-linux-2.37/sys-utils/lscpu-cputype.c	2021-07-04 15:09:29.174747744 +0200
@@ -6,6 +6,10 @@
 
 #include "fileutils.h"
 
+#if defined(HAVE_LIBRTAS)
+# include <librtas.h>
+#endif
+
 /* Lookup a pattern and get the value for format  "<pattern> : <key>"
  */
 int lookup(char *line, char *pattern, char **value)
