diff -ur util-linux-2.9r.orig/lib/pathnames.h util-linux-2.9r/lib/pathnames.h
--- util-linux-2.9r.orig/lib/pathnames.h	Sun Oct 11 21:19:16 1998
+++ util-linux-2.9r/lib/pathnames.h	Tue Jun  1 19:07:58 1999
@@ -64,8 +64,8 @@
 #define _PATH_TTY       	"/dev/tty"
 #define TTYTYPES        	"/etc/ttytype"
 #define SECURETTY       	"/etc/securetty"
-#define _PATH_UTMP      	"/var/run/utmp"
-#define _PATH_WTMP      	LOGDIR "/wtmp"
+#define _PATH_UTMP      	"/var/run/utmpx"
+#define _PATH_WTMP      	LOGDIR "/wtmpx"
 #define _PATH_WTMPLOCK		"/etc/wtmplock"
 
 /* no more . in DEFPATH */
Only in util-linux-2.9r/lib: pathnames.h~
