diff -ruNp util-linux-2.34.orig/bash-completion/umount util-linux-2.34/bash-completion/umount
--- util-linux-2.34.orig/bash-completion/umount	2019-04-02 12:12:03.180261025 +0200
+++ util-linux-2.34/bash-completion/umount	2019-11-24 23:13:33.844871532 +0100
@@ -51,7 +51,7 @@ _umount_module()
 
 	local oldifs=$IFS
 	IFS=$'\n'
-	COMPREPLY=( $( compgen -W "$(findmnt -lno TARGET | awk \
+	COMPREPLY=( $( compgen -W "$(findmnt -lno TARGET | gawk \
 		'{
 			if ($0 ~ ENVIRON["HOME"]) {
 				homeless = $0
