diff -urNp -x '*.orig' util-linux-2.36.2.org/include/pathnames.h util-linux-2.36.2/include/pathnames.h
--- util-linux-2.36.2.org/include/pathnames.h	2021-02-12 12:32:01.736988771 +0100
+++ util-linux-2.36.2/include/pathnames.h	2021-02-28 14:21:31.543044005 +0100
@@ -39,7 +39,7 @@
 #define _PATH_NOLOGIN_TXT	"/etc/nologin.txt"
 
 #ifndef _PATH_MAILDIR
-# define _PATH_MAILDIR		"/var/spool/mail"
+# define _PATH_MAILDIR		"/var/mail"
 #endif
 #define	_PATH_MOTDFILE		"/usr/share/misc/motd:/run/motd:/etc/motd"
 #ifndef _PATH_NOLOGIN
diff -urNp -x '*.orig' util-linux-2.36.2.org/login-utils/login.1 util-linux-2.36.2/login-utils/login.1
--- util-linux-2.36.2.org/login-utils/login.1	2021-02-12 12:32:01.748988728 +0100
+++ util-linux-2.36.2/login-utils/login.1	2021-02-28 14:21:31.543044005 +0100
@@ -347,7 +347,7 @@ the superuser logs in.  ENV_ROOTPATH tak
 .I /var/run/utmp
 .I /var/log/wtmp
 .I /var/log/lastlog
-.I /var/spool/mail/*
+.I /var/mail/*
 .I /etc/motd
 .I /etc/passwd
 .I /etc/nologin
