--- util-linux-2.30/include/pathnames.h.orig	2017-02-23 13:19:13.927424206 +0100
+++ util-linux-2.30/include/pathnames.h	2017-06-22 21:40:58.287666726 +0200
@@ -41,7 +41,7 @@
 #define _PATH_NOLOGIN_TXT	"/etc/nologin.txt"
 
 #ifndef _PATH_MAILDIR
-# define _PATH_MAILDIR		"/var/spool/mail"
+# define _PATH_MAILDIR		"/var/mail"
 #endif
 #define	_PATH_MOTDFILE		"/etc/motd"
 #ifndef _PATH_NOLOGIN
--- util-linux-2.21/login-utils/login.1~	2012-02-10 15:45:52.000000000 +0200
+++ util-linux-2.21/login-utils/login.1	2012-03-05 18:07:04.386917212 +0200
@@ -223,7 +223,7 @@
 .I /var/run/utmp
 .I /var/log/wtmp
 .I /var/log/lastlog
-.I /var/spool/mail/*
+.I /var/mail/*
 .I /etc/motd
 .I /etc/passwd
 .I /etc/nologin
