--- util-linux-2.10a/lib/pathnames.h.wiget	Fri Nov 12 00:24:49 1999
+++ util-linux-2.10a/lib/pathnames.h	Fri Nov 12 00:47:51 1999
@@ -86,6 +86,10 @@
 #define LOGDIR                  "/var/log"
 #endif
 
+#ifndef DATAMISCDIR
+#define DATAMISCDIR              "/usr/share/misc"
+#endif
+
 #ifndef VARPATH
 #define VARPATH			"/var"
 #endif
@@ -101,7 +105,7 @@
 #define SECURETTY       	"/etc/securetty"
 #define _PATH_UTMP      	"/var/run/utmpx"
 #define _PATH_WTMP      	LOGDIR "/wtmpx"
-#define _PATH_WTMPLOCK		"/etc/wtmplock"
+#define _PATH_WTMPLOCK		"/var/log/wtmpxlock"
 
 /* no more . in DEFPATH */
 #define	_PATH_DEFPATH	        "/usr/local/bin:/bin:/usr/bin"
@@ -110,7 +114,7 @@
 #define	_PATH_LASTLOG		LOGDIR "/lastlog"
 
 #ifndef _PATH_MAILDIR
-#define	_PATH_MAILDIR		VARPATH "/spool/mail"
+#define	_PATH_MAILDIR		VARPATH "/mail"
 #endif
 #define	_PATH_MOTDFILE		"/etc/motd"
 #define	_PATH_NOLOGIN		"/etc/nologin"
@@ -142,5 +146,5 @@
 #define _PATH_GTMPTMP           "/etc/gtmptmp"
 
 /* used in misc-utils/look.c */
-#define _PATH_WORDS             "/usr/dict/words"
-#define _PATH_WORDS_ALT         "/usr/dict/web2"
+#define _PATH_WORDS             "/usr/share/dict/words"
+#define _PATH_WORDS_ALT         "/usr/share/dict/web2"
